internal/venice-common/src/main/java/com/linkedin/venice/pubsub/adapter/PubSubSharedProducerAdapter.java

```diff
@@ -0,0 +1,178 @@
+import io.tehuti.metrics.MetricsRepository;
+import io.tehuti.metrics.Sensor;
+import io.tehuti.metrics.stats.Avg;
+import io.tehuti.metrics.stats.Max;
+import it.unimi.dsi.fastutil.objects.Object2DoubleMap;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Future;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * A wrapper with a unique identifier around a regular producer. This producer may be
+ * used to send data to multiple different topics.
+ */
+public class PubSubSharedProducerAdapter implements PubSubProducerAdapter {
+  private static final Logger LOGGER = LogManager.getLogger(PubSubSharedProducerAdapter.class);
+
+  private final PubSubSharedProducerFactory sharedProducerFactory;
+  private final Set<String> producerTasks = new HashSet<>();
+  private final Map<String, Double> producerMetrics = new HashMap<>();
+  private final PubSubProducerAdapter producerAdapter;
+  private final int id;
+  private long lastStatUpdateTsMs = 0;
+  private SharedProducerStats sharedProducerStats;
+
+  public PubSubSharedProducerAdapter(
+      PubSubSharedProducerFactory sharedProducerFactory,
+      PubSubProducerAdapter producerAdapter,
+      MetricsRepository metricsRepository,
+      Set<String> metricsToBeReported,
+      int id) {
+    this.sharedProducerFactory = sharedProducerFactory;
+    this.id = id;
+    this.producerAdapter = producerAdapter;
+    metricsToBeReported
+        .forEach(metric -> producerMetrics.put(metric, (double) StatsErrorCode.KAFKA_CLIENT_METRICS_DEFAULT.code));
+    if (metricsRepository != null && producerMetrics.size() > 0) {
+      sharedProducerStats = new SharedProducerStats(metricsRepository);
+    }
+  }
+
+  @Override
+  public int getNumberOfPartitions(String topic) {
+    return producerAdapter.getNumberOfPartitions(topic);
+  }
+
+  /**
+   * Sends a message to a topic using internal producer adapter.
+   * @param topic - The topic to be sent to.
+   * @param key - The key of the message to be sent.
+   * @param value - The {@link KafkaMessageEnvelope}, which acts as the value.
+   * @param callback - The callback function, which will be triggered when producer sends out the message.
+   * */
+  @Override
+  public Future<PubSubProduceResult> sendMessage(
+      String topic,
+      Integer partition,
+      KafkaKey key,
+      KafkaMessageEnvelope value,
+      PubSubMessageHeaders headers,
+      PubSubProducerCallback callback) {
+    long startNs = System.nanoTime();
+    Future<PubSubProduceResult> result = producerAdapter.sendMessage(topic, partition, key, value, headers, callback);
+    sharedProducerStats.recordProducerSendLatency(LatencyUtils.getLatencyInMS(startNs));
+    return result;
+  }
+
+  @Override
+  public void flush() {
+    producerAdapter.flush();
+  }
+
+  @Override
+  public void close(int closeTimeOutMs) {
+    producerAdapter.close(closeTimeOutMs);
```

Comment:
nit: can use the below method instead.  Or we can even remove this API as I see there are 4 close() APIs here...

Reply:
Done